Data refactor 2.0

tylerflex opened the pull request:
To Do List:

## Reorganize
- [x] separate `DataArray` from `xr.DataArray`
- [x] separate `Dataset` from `MonitorData`.
- [x] make changes to `SimulationData`.

## Basic Features
- [x] equality checking.
- [x] normalization
- [x] symmetry
- [x] colocation.
- [x] `SimulationData` plotting.
- [x] hdf5 file IO.
- [x] json file IO.
- [x] separate data file IO

## Propagate to dependents

- [x] Fix `smatrix`.
- [x] Fix `ModeSolver`.
- [x] Fix `near2far`.
- [x] Backend PR

## Usability

- [x] data selection from `SimulationData`.
- [x] equality checking between `DataArray`.
- [x] property linking `MonitorData` -> `xr.DataArray`

## Cleanup
- [x] pylint.
- [x] pass all tests.
- [x] make diagram for `tidy3d/components/data/README.md`
- [ ] test notebooks.

momchil-flex commented:
I just pushed a commit where:
- I wrapped `MonitorData.colocate` to call `self.dataset.colocate`
- Fixed normalization for near2far data
- Added near2far data to the test_sim_data tests

The last one of those breaks the tests already - something about encoding. I looked into it a little and it *seems* to come from writing of "dims", but I'm not sure.

There's also two more things that break the tests. Right now for me writing of simulation data to json seems broken, if the data is also to be written to json. In the json test, without the `extra_data` argument, my computer kills the job (runs OOM I believe?). In a backend test I did manage to write a sim_data to json, but reading it failed.

The other thing is in the hdf5 test. The mode amps data "direction" changes type during IO (becomes bytes string, which fails when you try to do e.g. `mode_data.amps.sel("+")`.) I think that also more generally the equivalence check should be upgraded to check both data array values and coordinates? Because this was missed by the test as it was before.

tylerflex commented:
Thanks, I rebased against develop with the recent near2far refactoring from #509 and then fixed all failing tests and a few more that were added.

The writing of simulation data or any other data component to json is especially problematic for two reasons:
1. it exports the whole `xr.DataArray` as a dictionary of lists for json compatibility, which can be very slow.
2. if the `dtype` of the `xr.DataArray` is `complex`, it will further store each of the data components as a dictionary `{'real': x, 'image': y}`, because complex values are not recognizable in json.

I added some handling so we **could** technically handle writing models containing data to json, but it was very slow so I had to make the test `SimulationData` much smaller so it would work. I wonder if we should just make external data files the standard..

tylerflex commented:
Thanks, there was in fact a bug in the equality checking.
I added a few more tests for that, and made sure to look at coords in the XR.DataArray equality check.
I also added a `DataArray.from_data_coords(data, coords)` class method constructor, which basically does

```py
DataArray.from_data_coords(data, coords)
```
is the same as
```py
DataArray.from_data_coords(data=xr.DataArray(data, coords=coords))
```
just to save a bit of time in the future.
